File: python/202_isHappy.py
# All non-happy numbers follow sequences that reach the cycle:
# 4, 16, 37, 58, 89, 145, 42, 20, 4, ... 

class Solution(object):
  def isHappy(self, n):
        """
        :type n: int
        :rtype: bool
        """
        Hash = {}

        while n != 1:
            if n in Hash: 
                return False
            else: 
                Hash[n] = True
            
            n = sum([pow(int(c) , 2) for c in str(n)])
        
        return True

File: C/202_isHappy.c
// If a number is NOT happy, there is always a 4 in the cycle.
bool isHappy(int n) {
    if (n <= 0)
        return false;
    
    int magic = 4;
    int t = 0;
    
    while (1) {
        if (n == 1) 
            return true;
        
        if (n == magic) 
            return false;

        while (n) {
            t += (n % 10) * (n % 10);
            n /= 10;
        }
        n = t;
    }
}
